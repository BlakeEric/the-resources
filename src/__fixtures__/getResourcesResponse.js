const getResourcesResponse = [
  {
    node: {
      id: 'd04da249-50e7-5b18-a992-d530e14831a7',
      frontmatter: {
        title: 'A List Apart',
        format: 'Article',
        focus: 'Front-end',
        skillLevel: ['Beginner','Moderate','Advanced']
      },
      fields: {
        slug: '/a-list-apart/'
      }
    }
  },
  {
    node: {
      id: '77a6be63-d3c7-50cb-bd29-438427560388',
      frontmatter: {
        title: 'Scrimba',
        format: 'Interactive',
        focus: 'Front-end',
        skillLevel: ['Beginner','Moderate']
      },
      fields: {
        slug: '/scrimba/'
      }
    }
  },
  {
    node: {
      id: '28feb8c5-acb2-5638-bdd6-b0457eff8db3',
      frontmatter: {
        title: 'Smashing Magazine',
        format: 'Article',
        focus: 'Multiple',
        skillLevel: ['Beginner','Moderate','Advanced']
      },
      fields: {
        slug: '/smashing-magazine/'
      }
    }
  },
  {
    node: {
      id: '744d8211-8c35-5c83-b8c1-ecb4ab48c0db',
      frontmatter: {
        title: 'The Odin Project',
        format: 'Tutorial',
        focus: 'Multiple',
        skillLevel: ['Beginner','Moderate']
      },
      fields: {
        slug: '/the-odin-project/'
      }
    }
  },
  {
    node: {
      id: '3b19247d-2ddd-558d-b0ce-14ec1972d52f',
      frontmatter: {
        title: 'Watch and Code',
        format: 'Video',
        focus: 'Front-end',
        skillLevel: ['Beginner']
      },
      fields: {
        slug: '/watch-and-code/'
      }
    }
  }
]

export default getResourcesResponse;
