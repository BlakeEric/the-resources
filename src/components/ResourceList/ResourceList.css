.table {
    border-collapse: collapse;
}
.table tbody tr {
    border-bottom: 1px solid #eaeaea; 
}