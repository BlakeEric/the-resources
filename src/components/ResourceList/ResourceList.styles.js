import { css } from '@emotion/core'
//import { rhythm } from "../../utils/typography"

export const resourceListStyles = css`
overflow-x: scroll;
border-radius: 5px;
padding: 1rem;
boxShadow: 0 2px 2px 0 rgba(0,0,0,0.05);
border: 1px solid #eaeaea;
table {
  min-width: 600px;
  position: relative;
}
tbody > tr {
  border-bottom: 1px solid #eaeaea;
}
td {
  width: 25%;
  padding-top: 0.25rem;
}
td span.notFoundMessage {
  width: 100%;
  text-align: center;
  display: inline-block;
}
td a {
  display: inline-block;
  line-height: 1.25;
}
`