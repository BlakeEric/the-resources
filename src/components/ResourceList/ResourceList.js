import React, { useContext } from "react"
import { resourceListStyles } from './ResourceList.styles.js'
import "./ResourceList.css"
import ResourceListHeader from '../ResourceListHeader/ResourceListHeader'
import Resource from '../Resource/Resource'
import { ResourceContext } from '../ResourceContext/ResourceContext'


export default function ResourceList() {

  const context = useContext(ResourceContext);
  const { resources } = context;
  const filteredResources = context.getFilteredResources();

  if (resources.length < 1) {
    return (
      <section id="resources">
        <h4>Something went wrong :(</h4>
        <p>We are unable to retrieve resources at this time.</p>
      </section>
    )
  }

  return (
    <div>
      <section className="tableWrap" css={resourceListStyles}>
        <table className="table">
          <ResourceListHeader keys={['name', 'format', 'focus', 'skillLevel']} />
          <tbody>
            {filteredResources.length === 0 ?
              <tr>
                <td colSpan="4">
                  <span className="notFoundMessage">Nothing found matching your search.</span>
                </td>
              </tr>
            : filteredResources.map(({ node }, i) => (
              <Resource
                key={node.id}
                {...node}
                showDetailsAbove={i > filteredResources.length - 5}
              />
            ))}
          </tbody>
        </table>
      </section>
    </div>
  )

}


ResourceList.defaultProps = {
  resources: [],
  skillLevels: [],
  formats:[],
  focuses: []
}

