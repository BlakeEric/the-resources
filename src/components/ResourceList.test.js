import React from "react"
import renderer from "react-test-renderer"
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import getResourcesResponse from '../__fixtures__/getResourcesResponse'

import ResourceList from "./ResourceList"

const mockProps = {
  resources: getResourcesResponse,
  skillLevels: ["Beginner", "Moderate", "Advanced"],
  focuses: ["Front-end", "Multiple", "Algorithms"],
  formats: ['Article', 'Interactive', 'Tutorial', 'Video']
}

describe("ResourceList", () => {

  /*
  * Make sure the component renders without error
  */
  describe('Initial render', () => {
    it("renders correctly when no resources passed in", () => {
      const tree = renderer
        .create(<ResourceList />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it("renders correctly when resources passed in", () => {
      const tree = renderer
        .create(<ResourceList {...mockProps} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it("orders ASC by name on initial render", () => {
      const wrapper = mount(<ResourceList {...mockProps} />)

      expect(wrapper.find('tbody tr').first().text().includes('Alligator')).toBeTruthy();
    });
  });


  /*
  * Sorting by Name
  */
  describe('When sorting by name', () => {
    it("toggles between DESC and ASC ordering by NAME on when appropriate button is clicked", () => {
      const wrapper = mount(<ResourceList {...mockProps} />)

      expect(wrapper.find('tbody tr').first().text().includes('Alligator')).toBeTruthy();

      // The "name" column button
      const button = wrapper.find('button').at(0)

      //trigger button click
      button.simulate('click');
      wrapper.update();

      // The last item's name should start with "A"
      expect(wrapper.find('tbody tr').last().text().includes('Alligator')).toBeTruthy();

      //trigger button click again
      button.simulate('click');
      wrapper.update();

      // The FIRST item's name should start with "A"
      expect(wrapper.find('tbody tr').first().text().includes('Alligator')).toBeTruthy();
    });
  });

  /*
  * Sorting by Format
  */
  describe('When sorting by Format', () => {
    it("toggles between DESC and ASC ordering by FORMAT on when appropriate button is clicked", () => {
      const wrapper = mount(<ResourceList {...mockProps} />)

      expect(wrapper.find('tbody tr').first().text().includes('Alligator')).toBeTruthy();

      // The "skillLevel" column button
      const button = wrapper.find('button').at(1)

      //trigger button click
      button.simulate('click');
      wrapper.update();

      // The first item's format should start with "A"
      expect(wrapper.find('tbody tr').first().text().includes('Article')).toBeTruthy();

      //trigger button click again
      button.simulate('click');
      wrapper.update();

      // The LAST item's format should start with "A"
      expect(wrapper.find('tbody tr').last().text().includes('Article')).toBeTruthy();
    });
  });


  /*
  * Sorting by Focus
  */
  describe('When sorting by Focus', () => {
    it("toggles between DESC and ASC ordering by FOCUS on when appropriate button is clicked", () => {
      const wrapper = mount(<ResourceList {...mockProps} />)

      expect(wrapper.find('tbody tr').first().text().includes('Alligator')).toBeTruthy();

      // The "focus" column button
      const button = wrapper.find('button').at(2)

      //trigger button click
      button.simulate('click');
      wrapper.update();

      // The first item's focus should start with "A"
      expect(wrapper.find('tbody tr').first().text().includes('Algorithms')).toBeTruthy();

      //trigger button click again
      button.simulate('click');
      wrapper.update();

      // The LAST item's focus should start with "A"
      expect(wrapper.find('tbody tr').last().text().includes('Algorithms')).toBeTruthy();
    });
  });

  /*
  * Sorting by skillLevel
  */
  describe('When sorting by Skill Level', () => {
    it("toggles between DESC and ASC ordering by SKILL LEVEL on when appropriate button is clicked", () => {
      const wrapper = mount(<ResourceList {...mockProps} />)

      expect(wrapper.find('tbody tr').first().text().includes('Alligator')).toBeTruthy();

      // The "skillLevel" column button
      const button = wrapper.find('button').at(3)

      //trigger button click
      button.simulate('click');
      wrapper.update();

      // The first item's skill level should start with "A"
      expect(wrapper.find('tbody tr').first().text().includes('Advanced')).toBeTruthy();

      //trigger button click again
      button.simulate('click');
      wrapper.update();

      // The LAST item's skill level should start with "A"
      expect(wrapper.find('tbody tr').last().text().includes('Advanced')).toBeTruthy();
    });
  });


  /*
  * Filtering by Name
  */
  describe('When filtering by name', () => {
    it("displays only items with Names containing the search term", () => {
      const wrapper = mount(<ResourceList {...mockProps} />)

      // Five items should be rendered
      expect(wrapper.find('tbody tr').length).toEqual(5);

      // The "search term" input
      const field = wrapper.find('input#searchTerm').first()

      field.instance().value = "Elephant";
      field.simulate('change');
      wrapper.update();

      // Only two items should remain after filtering
      expect(wrapper.find('tbody tr').length).toEqual(1);
    });
  });

  /*
  * Filtering by skillLevel
  */
  describe('When filtering by Skill Level', () => {
    it("displays only items with skill levels containing the selection", () => {
      const wrapper = mount(<ResourceList {...mockProps} />)

      // Five items should be rendered
      expect(wrapper.find('tbody tr').length).toEqual(5);

      // The "skillLevel" filter button
      const field = wrapper.find('select#resourceSkillLevel').first()

      field.instance().value = "Advanced";
      field.simulate('change');
      wrapper.update();

      // Only two items should remain after filtering
      expect(wrapper.find('tbody tr').length).toEqual(2);
    });
  });

  /*
  * Filtering by format
  */
  describe('When filtering by Format', () => {
    it("displays only items with formats containing the selection", () => {
      const wrapper = mount(<ResourceList {...mockProps} />)

      // Five items should be rendered
      expect(wrapper.find('tbody tr').length).toEqual(5);

      // The "format" filter button
      const field = wrapper.find('select#resourceFormat').first()

      field.instance().value = "Tutorial";
      field.simulate('change');
      wrapper.update();

      // Only two items should remain after filtering
      expect(wrapper.find('tbody tr').length).toEqual(1);
    });
  });

  /*
  * Filtering by focus
  */
  describe('When filtering by Focus', () => {
    it("displays only items with formats containing the selection", () => {
      const wrapper = mount(<ResourceList {...mockProps} />)

      // Five items should be rendered
      expect(wrapper.find('tbody tr').length).toEqual(5);

      // The "focus" filter button
      const field = wrapper.find('select#resourceFocus').first()

      field.instance().value = "Front-end";
      field.simulate('change');
      wrapper.update();

      // Only two items should remain after filtering
      expect(wrapper.find('tbody tr').length).toEqual(2);
    });
  });


  /*
  * Clearing filters
  */
  describe('When filtering by Focus', () => {
    it("displays all items when 'Clear filter' button is clicked", () => {

      const wrapper = mount(<ResourceList {...mockProps} />)

      // The "skillLevel" filter button
      const field = wrapper.find('select#resourceFocus').first()

      field.instance().value = "Front-end";
      field.simulate('change');
      wrapper.update();

      // Only two items should remain after filtering
      expect(wrapper.find('tbody tr').length).toEqual(2);

      // Simulate "clear filters" click
      const button = wrapper.find('.btn-clearFilters').first()
      button.simulate('click');
      wrapper.update();

      // All items should be showing again
      expect(wrapper.find('tbody tr').length).toEqual(5);

    });
  });


})
