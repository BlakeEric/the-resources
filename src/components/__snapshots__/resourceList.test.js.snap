// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceList Initial render renders correctly when no resources passed in 1`] = `
<section
  id="resources"
>
  <h4>
    Something went wrong :(
  </h4>
  <p>
    We are unable to retrieve resources at this time.
  </p>
</section>
`;

exports[`ResourceList Initial render renders correctly when resources passed in 1`] = `
<div>
  <form>
    <h3>
      Filter!
    </h3>
    <label
      htmlFor="searchTerm"
    >
      Name
      <input
        aria-label="Name"
        id="searchTerm"
        name="searchTerm"
        onChange={[Function]}
        placeholder="search resources"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="resourceSkillLevel"
    >
      Skill Level
    </label>
    <select
      id="resourceSkillLevel"
      name="resourceSkillLevel"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        All
      </option>
      <option
        value="Beginner"
      >
        Beginner
      </option>
      <option
        value="Moderate"
      >
        Moderate
      </option>
      <option
        value="Advanced"
      >
        Advanced
      </option>
    </select>
    <label
      htmlFor="resourceFormat"
    >
      Format
    </label>
    <select
      id="resourceFormat"
      name="resourceFormat"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        All
      </option>
      <option
        value="Article"
      >
        Article
      </option>
      <option
        value="Interactive"
      >
        Interactive
      </option>
      <option
        value="Tutorial"
      >
        Tutorial
      </option>
      <option
        value="Video"
      >
        Video
      </option>
    </select>
    <label
      htmlFor="resourceFocus"
    >
      Focus
    </label>
    <select
      id="resourceFocus"
      name="resourceFocus"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        All
      </option>
      <option
        value="Front-end"
      >
        Front-end
      </option>
      <option
        value="Multiple"
      >
        Multiple
      </option>
      <option
        value="Algorithms"
      >
        Algorithms
      </option>
    </select>
  </form>
  <table
    id="resources"
  >
    <thead>
      <tr>
        <th
          className="active"
        >
          <button
            onClick={[Function]}
          >
            Name
             
            â†“
          </button>
        </th>
        <th
          className=""
        >
          <button
            onClick={[Function]}
          >
            Format
             
            
          </button>
        </th>
        <th
          className=""
        >
          <button
            onClick={[Function]}
          >
            Focus
             
            
          </button>
        </th>
        <th
          className=""
        >
          <button
            onClick={[Function]}
          >
            Skill Level
             
            
          </button>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <a
            href="/alligator/"
          >
            Alligator
          </a>
        </td>
        <td>
          Article
        </td>
        <td>
          Front-end
        </td>
        <td>
          All
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/badger/"
          >
            Badger
          </a>
        </td>
        <td>
          Interactive
        </td>
        <td>
          Algorithms
        </td>
        <td>
          Moderate
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/caterpillar/"
          >
            Caterpillar
          </a>
        </td>
        <td>
          Article
        </td>
        <td>
          Multiple
        </td>
        <td>
          Advanced
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/dolphin/"
          >
            Dolphin
          </a>
        </td>
        <td>
          Tutorial
        </td>
        <td>
          Multiple
        </td>
        <td>
          Beginner, Moderate
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/elephant/"
          >
            Elephant
          </a>
        </td>
        <td>
          Video
        </td>
        <td>
          Front-end
        </td>
        <td>
          Beginner
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
